pydantic/type_adapter.py

```diff
@@ -233,100 +208,129 @@ def __init__(
         self._type = type
         self._config = config
         self._parent_depth = _parent_depth
+
         if module is None:
             f = sys._getframe(1)
             self._module_name = cast(str, f.f_globals.get('__name__', ''))
         else:
             self._module_name = module
 
-        self._core_schema: CoreSchema | None = None
-        self._validator: SchemaValidator | PluggableSchemaValidator | None = None
-        self._serializer: SchemaSerializer | None = None
+        self.core_schema: CoreSchema
+        self.validator: SchemaValidator | PluggableSchemaValidator
+        self.serializer: SchemaSerializer
+
+        localns: _namespace_utils.MappingNamespace = (
+            _typing_extra.parent_frame_namespace(parent_depth=self._parent_depth) or {}
+        )
+        globalns: _namespace_utils.GlobalsNamespace = sys._getframe(max(self._parent_depth - 1, 1)).f_globals.copy()
+        ns_resolver = _namespace_utils.NsResolver(
+            namespaces_tuple=_namespace_utils.NamespacesTuple(locals=localns, globals=globalns),
+            parent_namespace=localns,
+        )
+        self._init_core_attrs(ns_resolver=ns_resolver, force=False)
+
+    def _init_core_attrs(self, ns_resolver: _namespace_utils.NsResolver, force: bool) -> bool:
+        """Initialize the core schema, validator, and serializer for the type.
 
-        if not self._defer_build():
-            # Immediately initialize the core schema, validator and serializer
-            with self._with_frame_depth(1):  # +1 frame depth for this __init__
-                # Model itself may be using deferred building. For backward compatibility we don't rebuild model mocks
-                # here as part of __init__ even though TypeAdapter itself is not using deferred building.
-                self._init_core_attrs(rebuild_mocks=False)
+        If `force` is set to `False` and `_defer_build` is `True`, the core schema, validator, and serializer will be set to mocks.
 
-    @contextmanager
-    def _with_frame_depth(self, depth: int) -> Iterator[None]:
-        self._parent_depth += depth
-        try:
-            yield
-        finally:
-            self._parent_depth -= depth
+        Args:
+            ns_resolver: The namespace resolver to use when building the core schema for the adapted type.
+            force: Whether to force the initialization of the core schema, validator, and serializer.
+
+        Returns:
+            `True` if the core schema, validator, and serializer were successfully initialized, otherwise `False`.
+        """
+        if not force and self._defer_build:
+            _mock_val_ser.set_type_adapter_mocks(self, str(self._type))
```

Comment:
Not sure is there any gains but the type stringification could be also lazy as its only used for errors really 🤔 So depends on how complex implementation there can be for the type stringifications